src/TestFramework/TestFramework/Assertions/Assert.AreSame.cs

```diff
@@ -10,6 +12,74 @@ namespace Microsoft.VisualStudio.TestTools.UnitTesting;
 /// </summary>
 public sealed partial class Assert
 {
+    [InterpolatedStringHandler]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public readonly struct AssertAreSameInterpolatedStringHandler<TArgument>
+    {
+        private readonly StringBuilder? _builder;
+        private readonly TArgument? _expected;
+        private readonly TArgument? _actual;
+
+        public AssertAreSameInterpolatedStringHandler(int literalLength, int formattedCount, TArgument? expected, TArgument? actual, out bool shouldAppend)
+        {
+            _expected = expected;
+            _actual = actual;
+            shouldAppend = IsAreSameFailing(expected, actual);
+            if (shouldAppend)
+            {
+                _builder = new StringBuilder(literalLength + formattedCount);
+            }
+        }
+
+        internal void FailIfNeeded()
+        {
+            if (_builder is not null)
+            {
+                FailAreSame(_expected, _actual, _builder.ToString());
+            }
+        }
+
+        public void AppendLiteral(string value) => _builder!.Append(value);
+
+        public void AppendFormatted<T>(T value) => _builder!.Append(value);
+
+#if NETCOREAPP3_1_OR_GREATER
+        public void AppendFormatted(ReadOnlySpan<char> value) => _builder!.Append(value);
```

Comment:
We may have problems if people are using ROS in .NET Framework via NuGet package. Not sure if it's acceptable to tell them "call `.ToString()`", or if we should take an extra dependency for netstandard/netframework